File: Algorithms/src/1458. Max Dot Product of Two Subsequences/Code/solution.cpp
class Solution {
public:
    int maxDotProduct(vector<int>& nums1, vector<int>& nums2) {
        int n = nums1.size();
        int m = nums2.size();
        
        // dp[i][j] = max dot product using nums1[i:] and nums2[j:]
        vector<vector<int>> dp(n + 1, vector<int>(m + 1, INT_MIN));
        
        // Fill DP table from bottom-right to top-left
        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                int product = nums1[i] * nums2[j];
                
                // Option 1: take both elements
                int takeBoth = product;
                if (dp[i + 1][j + 1] != INT_MIN)
                    takeBoth = max(takeBoth, product + dp[i + 1][j + 1]);
                
                // Option 2 & 3: skip one element
                dp[i][j] = max({
                    takeBoth,
                    dp[i + 1][j],
                    dp[i][j + 1]
                });
            }
        }
        
        return dp[0][0];
    }
};

File: Algorithms/src/1458. Max Dot Product of Two Subsequences/Code/solution.py
class Solution:
    def maxDotProduct(self, nums1: List[int], nums2: List[int]) -> int:
        n, m = len(nums1), len(nums2)
        
        dp = [[float('-inf')] * (m + 1) for _ in range(n + 1)]
        
        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                product = nums1[i] * nums2[j]
                
                take_both = product
                if dp[i + 1][j + 1] != float('-inf'):
                    take_both = max(take_both, product + dp[i + 1][j + 1])
                
                dp[i][j] = max(
                    take_both,
                    dp[i + 1][j],
                    dp[i][j + 1]
                )
        
        return dp[0][0]
